### ISSUE TITLE
New rule proposal: no-dupe-else-if

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12469]
**Please describe what the rule should do:**

Disallows duplicate test conditions in `if-else-if` chains.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)

**Provide 2-3 code examples that this rule will warn about:**

```js
if (a > 1) {
    foo();
} else if (a > 1) {
    bar();
}

if (a === 1) {
    foo();
} else if (a === 2) {
    bar();
} else if (a === 3) { 
    baz();
} else if (a === 2) {
    quux();
} else if (a === 5) {
    quuux();
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a possible error. A statement with a duplicate condition can never execute (unless it's an unusual use case with side effects).

This rule is similar to the existing `no-duplicate-case`. The error could be due to a copy & paste and a long chain.

The test nodes would be compared by their tokens.

**Are you willing to submit a pull request to implement this rule?**

Yes.

**Comment by mdjermanovic** [comment]
One more :+1: to accept this.

**Comment by mysticatea** [comment]
Now accepted :tada:

**Comment by mdjermanovic** [comment]
I'm working on this :-)

**Comment by aladdin-add** [comment]
just a thought: for consecutive `if`, whether or not it should check, e.g.
```js
if(foo === 0) {...}
if(foo === 0) {...} // to warn, or not?
```

I think it's good to cover it, or behide an option as an enhancement.

**Comment by mdjermanovic** [comment]
> just a thought: for consecutive `if`, whether or not it should check, e.g.
> 
> ```js
> if(foo === 0) {...}
> if(foo === 0) {...} // to warn, or not?
> ```
> 
> I think it's good to cover it, or behide an option as an enhancement.

I think it's more suitable for a separate rule because this isn't necessarily an error (it could be just a bad practice to repeat conditions instead of having all code in a single block, or it might be even a correct code if `foo` can be assigned a different value in the first block), unlike certainly unreachable code that's being reported by the rule (assuming there are no side effects in conditions).

**Comment by kaicataldo** [comment]
> just a thought: for consecutive `if`, whether or not it should check, e.g.
> ```js
> if(foo === 0) {...}
> if(foo === 0) {...} // to warn, or not?
> ```
>
> I think it's good to cover it, or behide an option as an enhancement.

Agreed with @mdjermanovic. I don't think this rule should catch this, since this rule shouldn't be disallowing code like the following:

```js
function checkForB(a) { return a.b; }
function checkForC(a) { return a.c; }

const d = {
 c: true
};

let e = checkForB(d);
if (e !== undefined) {/* Do something */}
e = checkForC(d);
if (e !== undefined) {/* Do something */}
```

or even if we limit it to consecutive statements, potentially:


```js
function checkForB(a) { return a.b; }
function checkForC(a) { return a.c; }

const d = {
 c: true
};

let e = checkForB(d);
if (e !== undefined) {
  if (typeof e !== 'number') {
    e = checkForC(d);
  } 
}
if (e !== undefined) {/* Do something */}
```

**Comment by aladdin-add** [comment]
emm.. after thinking for a while, I think it's not fit for this rule:
+ its rulename: no-dupe-else-if :)
+ it can catch some mistakes, but also has a few false positives.
